design/BufferOperations.md

```diff
@@ -0,0 +1,150 @@
+On that boundary:
+
+ - The `WebGPUMappedMemory` goes in the available state.
+ - If the `WebGPUMappedMemory` was created via `WebGPUBuffer.mapWrite`, its content is cleared to 0.
+ - `success` is called with the content of the memory as an argument.
+
+If `success` hasn't been called when the WebGPUMappedMemory gets invalidated (meaning the object is still in the pending state), `error` is called instead. When `WebGPUMappedMemory` goes from the available state to the invalidated state, the `ArrayBuffer` for its content gets neutered. The return value of `then` acts like the return value of `Promise.then`.
+
+The `ArrayBuffer` of a `WebGPUMappedMemory` created from a `mapWrite` is where the application should write the data and its content is made available to the buffer when the `WebGPUMappedMemory` is invalidated (i.e. `WebGPUBuffer.unmap` is called).
+
+## Immediate data upload
+
+Buffer mapping is the path with the least number of copies but it is often useful to upload data to a buffer *right now*, if only for debugging.
+A `WebGPUBuffer` operation is provided that takes an ArrayBuffer and copies its content at an offset in the buffer.
+
+```
+partial interface WebGPUBuffer {
+    void setSubData(ArrayBuffer data, u32 offset);
+}
+```
+
+This operation acts as if it was done after all previous "device-level" commands and before all subsequent "device-level" commands.
+"Device level" commands are all commands not buffered in a `WebGPUCommandBuffer`, and include `WebGPUQueue.submit`.
+The content of `data` is only read during the call and can be modified by the application afterwards.
+The following must be true or a validation error occurs:
+
+ - The buffer must have been created with the `WebGPUBufferUsage.TRANSFER_DST` usage flag.
+ - `offset + data.length` must not overflow and be at most the size of the buffer.
+ - Depending on the design of memory barriers, the buffer must be, or allowed to be in the `WebGPUBufferUsage.TRANSFER_DST` usage.
+   - In particular the buffer must not be currently mapped.
+
+## Unused designs
+
+### Persistently mapped buffer
+
+Persistently mapped buffer are when the result of mapping the buffer can be kept by the application while the buffer is in use by the GPU.
+We didn't find a way to have persistently mapped buffers and at the same time keep things data race free between the CPU and GPU.
+Being data race free would be possible if ArrayBuffer could be unneutered but this is not the case.
+
+### Promise<ArrayBuffer> readback();
+
+This didn't have a pollable interface and forced an extra buffer-to-buffer copy to occur if the GPU execution could be resumed immediately.
+
+### NXT's MapReadAsync(callback);
+
+Not a pollable interface.
+
+## Issues
+
+### GC discoverability
+
+It isn't clear yet what happens when a buffer gets garbage collected while it is mapped.
+The simple answer is that the `WebGPUMappedMemory` objects get invalidated but that would allow the application to discover when the GC runs.
```

Comment:
Wouldn't the simpler answer be "don't collect the object?" Either by bumping a reference count upon map, or by pinning the object?

Reply:
Good point, don't know why I didn't think of it.
I think the WebGPUMappedMemory should point to the buffer. It could even be in a public attribute. (Makes it easier to do the unmap operation on one-off mappings.

Reply:
Having buffer as an attribute of the WebGPUMappedMemory would work and is pretty simple :+1: